.changesets/feat_garypen_2002_subgraph_batching.md

```diff
@@ -0,0 +1,27 @@
+### Subgraph support for query batching ([Issue #2002](https://github.com/apollographql/router/issues/2002))
+
+This project is an extension of the existing work to support [client side batching in the router](https://github.com/apollographql/router/issues/126). The current implementation is experimental and is publicly [documented](https://www.apollographql.com/docs/router/executing-operations/query-batching/).
+
+Currently the concept of a batch is preserved until the end of the `RouterRequest` processing. At this point, we convert each batch request item into a separate `SupergraphRequest`. These are then planned and executed concurrently within the router and re-assembled into a batch when they complete. It's important to note that, with this implementation, the concept of a batch, from the perspective of an executing router, now disappears and each request is planned and executed separately.
+
+This extension will modify the router so that the concept of a batch is preserved, at least outwardly, so that multiple subgraph requests are "batched" (in exactly the same format as a client batch request) for onward transmission to subgraphs. The goal of this work is to provide an optimisation by reducing the number of round-trips to a subgraph from the router.
+
+Illustrative configuration.
+
+```yaml
+batching:
+  enabled: true
+  mode: batch_http_link
+  subgraph:
+    all:
+      enabled: true
+    subgraphs:
+      subgraph_1:
+        enabled: true
+      subgraph_2:
+        enabled:true
+````
```

Comment:
This is invalid to have both `all` and `subgraphs`, correct? If yes, should replace with separate configs for the two. Also, what's logged/returned by the router if this invalid config is set?

Reply:
I split the config sample into 2 so it was valid. If you did try to use this configuration, you'd see something like:
```
2024-04-03T08:48:41.981376Z ERROR  Failed to read configuration: configuration had errors: 
1. at line 13

  
  batching:
    enabled: true
    mode: batch_http_link
┌   subgraph:
|     all:
|       enabled: true
|     subgraphs:
|       subgraph_1:
|         enabled: true
|       subgraph_2:
|         enabled:true
└-----> {"all":{"enabled":true},"subgraphs":{"subgraph_1":{"enabled":true},"subgraph_2":"enabled:true"}} is not valid under any of the schemas listed in the 'oneOf' keyword


2024-04-03T08:48:41.981728Z INFO  stopped
2024-04-03T08:48:41.981984Z ERROR  no valid configuration was supplied
no valid configuration was supplied
```
which is the usual kind of error your would see for an "all" vs "specific" in router config.

fixed: https://github.com/apollographql/router/pull/4661/commits/7c138693a2f713ee9c0c404dbacc53f6a02cec30